lgc/builder/YCbCrConverter.cpp

```diff
@@ -326,64 +328,78 @@ void YCbCrConverter::genSamplerDescChroma() {
-  Value *heightHalf = m_builder->CreateLShr(height, pInt32One);
   m_width = m_builder->CreateUIToFP(width, m_builder->getFloatTy());
   m_height = m_builder->CreateUIToFP(height, m_builder->getFloatTy());
 
-  Value *pPitch = proxySqRsrcRegHelper.getReg(SqRsrcRegs::Pitch);
-  Value *pitchHalf = m_builder->CreateLShr(pPitch, pInt32One);
+  if (m_metaData.word1.planes == 1) {
+    Value *imgDataFmt = proxySqRsrcRegHelper.getReg(SqRsrcRegs::Format);
+    Value *dstSelXYZW = proxySqRsrcRegHelper.getReg(SqRsrcRegs::DstSelXYZW);
+    Value *isGBGRFmt = nullptr;
+    Value *isBGRGFmt = nullptr;
+
+    switch (m_gfxIp->major) {
+    case 9: {
+      isGBGRFmt = m_builder->CreateICmpEQ(
+          imgDataFmt, m_builder->getInt32(Pal::Gfx9::Chip::IMG_DATA_FORMAT::IMG_DATA_FORMAT_BG_RG__CORE));
+
+      isBGRGFmt = m_builder->CreateICmpEQ(
+          imgDataFmt, m_builder->getInt32(Pal::Gfx9::Chip::IMG_DATA_FORMAT::IMG_DATA_FORMAT_GB_GR__CORE));
+
+      proxySqRsrcRegHelper.setReg(SqRsrcRegs::Format,
+                                  m_builder->getInt32(Pal::Gfx9::Chip::IMG_DATA_FORMAT::IMG_DATA_FORMAT_8_8_8_8));
+      break;
+    }
+    case 10: {
+      isGBGRFmt = m_builder->CreateICmpEQ(
+          imgDataFmt, m_builder->getInt32(Pal::Gfx9::Chip::IMG_FMT::IMG_FMT_BG_RG_UNORM__GFX10CORE));
+      isBGRGFmt = m_builder->CreateICmpEQ(
+          imgDataFmt, m_builder->getInt32(Pal::Gfx9::Chip::IMG_FMT::IMG_FMT_GB_GR_UNORM__GFX10CORE));
+
+      proxySqRsrcRegHelper.setReg(SqRsrcRegs::Format,
+                                  m_builder->getInt32(Pal::Gfx9::Chip::IMG_FMT::IMG_FMT_8_8_8_8_UNORM__GFX10CORE));
+      break;
+    }
+    default:
+      llvm_unreachable("GFX IP not supported!");
+      break;
+    }
 
-  proxySqRsrcRegHelper.setReg(SqRsrcRegs::Width, widthHalf);
-  proxySqRsrcRegHelper.setReg(SqRsrcRegs::Format, m_builder->getInt32(m_metaData.word3.sqImgRsrcWord1));
+    dstSelXYZW = m_builder->CreateSelect(isGBGRFmt, m_builder->getInt32(0x977), dstSelXYZW);
+    dstSelXYZW = m_builder->CreateSelect(isBGRGFmt, m_builder->getInt32(0xF2E), dstSelXYZW);
 
-  switch (m_metaData.word1.planes) {
-  case 1:
-    proxySqRsrcRegHelper.setReg(SqRsrcRegs::DstSelXYZW, m_builder->getInt32(m_metaData.word1.dstSelXYZW));
-    proxySqRsrcRegHelper.setReg(SqRsrcRegs::BcSwizzle, m_builder->getInt32(6));
+    YCbCrAddressHandler addrHelper(m_builder, &proxySqRsrcRegHelper, m_gfxIp);
+    addrHelper.genHeightAndPitch(m_metaData.word0.bitDepth.channelBitsR, 32, m_metaData.word2.bitCounts.xBitCount,
+                                 m_metaData.word1.tileOptimal, m_metaData.word1.planes);
+    proxySqRsrcRegHelper.setReg(SqRsrcRegs::Width,
+                                m_builder->CreateLShr(width, ConstantInt::get(m_builder->getInt32Ty(), 1)));
+    proxySqRsrcRegHelper.setReg(SqRsrcRegs::DstSelXYZW, dstSelXYZW);
     proxySqRsrcRegHelper.setReg(SqRsrcRegs::Pitch, addrHelper.getPitchCb());
     m_imgDescsChroma[1] = proxySqRsrcRegHelper.getRegister();
```

Comment:
Why are we setting `m_imgDescsChroma[1]` if there's only one plane? Shouldn't it be `m_imgDescsChroma[0]`?

Reply:
This is mentioned in the design note. When we encountered GbGr or BgRg format(Special one plane case), the m_imgDescsChroma[0] is used for sampling Y channel and the modified the descriptor is saved in m_imgDescsChroma[1] for sampling CbCr channel.